File: 15-3sum/3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()

        for i , a in enumerate(nums):
            if a > 0: # Skip positive
                break

            if i > 0 and nums[i-1] == a:
                continue
            
            p1 = i + 1
            p2 = len(nums) - 1

            while p1 < p2:
                k = a + nums[p1] + nums[p2]
                if k > 0:
                    p2 -= 1
                elif k < 0:
                    p1 += 1
                else:
                    res.append([a,nums[p1],nums[p2]])
                    p1 += 1
                    while nums[p1] == nums[p1-1] and p1 < p2:
                        p1 += 1
                    p2 -= 1
                    while nums[p2] == nums[p2+1] and p1 < p2:
                        p2 -= 1

        return res

File: 15-3sum/3sum.ts
function threeSum(nums: number[]): number[][] {
    nums.sort((a, b) => a - b);
    const n = nums.length;
    const res = [];
    const added = new Set<string>();
    if (nums[0] > 0 || nums[n - 1] < 0) return res;
    for (let i = 0; i < n - 2; i++) {
        if (i > 0 && nums[i] === nums[i-1]) continue;

        let left = i + 1;
        let right = n - 1;

        while (left < right) {
            const sum = nums[i] + nums[left] + nums[right];
            if (sum === 0) {
                res.push([nums[i], nums[left], nums[right]]);

                while (left < right && nums[left] === nums[left+1]) left++;
                while (left < right && nums[right-1] === nums[right]) right--;

                left++;
                right--;
            }
            else if (sum < 0) {
                left++;
            }
            else {
                right--;
            }
        }
    }
    return res;
};